### PR TITLE
Use constants instead of hardcoded Requirement-Fieldname strings.

### PR DESCRIPTION
**Pull request by GGBeer**
Separated part 1 Replacing strings.
See Discussions #741 GrammarElementField-TypeValue and External-Refs

Thx,
Guus

### CODE REVIEW

**Comment by stanislaw** [comment]
You need to run `invoke check` locally and resolve all the issues because that's what the CI server runs as well.

Furthermore, regarding the commit name, as suggested in the https://strictdoc.readthedocs.io/en/latest/strictdoc-6-contributing.html, please add the context something like `sdoc/models: Use constants instead...`.

**Comment by GGBeer** [comment]
Hi, sorry for the extra work.

I found one issue which needs discussion.
The unit test tests/unit/strictdoc/backend/sdoc/test_dsl_passthrough.py test_100_basic_test() fails during grammar validation.
Seems that this failure is unrelated to my changes and must be failing for a while.

The test contains a TAGS field:
` "TAGS: Tag 1, Tag 2, Tag 3"`

This test is broken by the verification in:

strictdoc/backend/sdoc/validations/requirement.py
```
def validate_requirement_field(
line:153..
    elif isinstance(grammar_field, GrammarElementFieldTag):
        requirement_field_value = requirement_field.field_value

        if not multi_choice_regex_match(requirement_field_value):

=>
def multi_choice_regex_match(value):
    keyword = "[a-zA-Z0-9_]+"
    regex = re.compile(rf"^{keyword}(, {keyword})*$")
```
This RegExpr does not allow spaces or any other non-AlphaNum characters beside underscores in Tags.

This multi_choice_regex_match was added by the commit
`Stanislav Pankevich 24.12.2021 16:28 backend/dsl: support grammar fields: SingleChoice, MultipleChoice, Tag`

Either the use of this RegExpr for GrammarElementFieldTag fields has to be changed or the Tags have to be restricted to AlphaNum+Underscore i.e. updating the unit test and documentation accordingly.

I would prefer a specific GrammarElementFieldTag-RegExpr allowing some additional punctuation characters like underscore, hyphen, collon, period, slash.  
What is your opinion here?

Thx,
Guus

**Comment by stanislaw** [comment]
This test has been passing for me locally and on the CI all the time. Are you sure you are developing on top of a non-modified `main` branch?

```
/Users/Stanislaw/.MyPython/Python_3.6.15/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py --target test_dsl_passthrough.py::test_100_basic_test 
Testing started at 18:27 ...
Launching pytest with arguments test_dsl_passthrough.py::test_100_basic_test --no-header --no-summary -q in /Users/Stanislaw/workspace/projects/code/strictdoc/tests/unit/strictdoc/backend/sdoc

============================= test session starts ==============================
collecting ... collected 1 item

test_dsl_passthrough.py::test_100_basic_test 

============================== 1 passed in 0.19s ===============================

Process finished with exit code 0
PASSED                      [100%]
```

With this result, I assume that your change should not have introduced any change in the functionality, but you only did the refactoring of the string literals. The errors that you are seeing should be the result of some other interference.

**Comment by GGBeer** [comment]
Strange, I am pretty sure that I branched from the main branch.
Maybe it's a local PyCharm/Python 3.10 version thing. 
However, by analyzing the code, the test should currently fail due to the spaces included into the Tag-Values.

```
C:\Workspace\GitHub\strictdoc\strictdoc.use-constants\venv\Scripts\python.exe "C:/Program Files/JetBrains/PyCharm Community Edition 2022.2.2/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py" --target test_dsl_passthrough.py::test_100_basic_test 
Testing started at 09:45 ...
Launching pytest with arguments test_dsl_passthrough.py::test_100_basic_test --no-header --no-summary -q in C:\Workspace\GitHub\strictdoc\strictdoc.use-constants\tests\unit\strictdoc\backend\sdoc

============================= test session starts =============================
collecting ... collected 1 item

test_dsl_passthrough.py::test_100_basic_test FAILED                      [100%]
tests\unit\strictdoc\backend\sdoc\test_dsl_passthrough.py:628 (test_100_basic_test)
...
...
>       document = reader.read(input_sdoc)

test_dsl_passthrough.py:661: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\..\..\..\strictdoc\backend\sdoc\reader.py:43: in read
    document, _ = SDReader._read(input_string, file_path)
..\..\..\..\..\strictdoc\backend\sdoc\reader.py:30: in _read
    document = meta_model.model_from_str(input_string, file_name=file_path)
..\..\..\..\..\venv\lib\site-packages\textx\metamodel.py:642: in model_from_str
    model = self._parser_blueprint.clone().get_model_from_str(
..\..\..\..\..\venv\lib\site-packages\textx\model.py:381: in get_model_from_str
    model = parse_tree_to_objgraph(
..\..\..\..\..\venv\lib\site-packages\textx\model.py:903: in parse_tree_to_objgraph
    call_obj_processors(m._tx_metamodel, m)
..\..\..\..\..\venv\lib\site-packages\textx\model.py:759: in call_obj_processors
    result = call_obj_processors(metamodel,
..\..\..\..\..\venv\lib\site-packages\textx\model.py:759: in call_obj_processors
    result = call_obj_processors(metamodel,
..\..\..\..\..\venv\lib\site-packages\textx\model.py:775: in call_obj_processors
    return_value_current = metamodel.process(
..\..\..\..\..\venv\lib\site-packages\textx\metamodel.py:520: in process
    return self._obj_processors.get(_type, lambda x: x)(value)
..\..\..\..\..\strictdoc\backend\sdoc\processor.py:162: in process_requirement
    validate_requirement(requirement, document_grammar)
..\..\..\..\..\strictdoc\backend\sdoc\validations\requirement.py:64: in validate_requirement
    valid_or_not_required_field = validate_requirement_field(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

requirement = Requirement(ng_level: None, level: None, uid: None, title_or_none: None, statement: System shall do X)
document_grammar = <strictdoc.backend.sdoc.models.document_grammar.DocumentGrammar object at 0x0000013F8DE138B0>
requirement_field = RequirementField(field_name: TAGS, field_value: Tag 1, Tag 2, Tag 3, field_value_multiline: None, field_value_references: [], )
grammar_field = <strictdoc.backend.sdoc.models.type_system.GrammarElementFieldTag object at 0x0000013F8DE138E0>
...
...
        elif isinstance(grammar_field, GrammarElementFieldTag):
            requirement_field_value = requirement_field.field_value
    
            if not multi_choice_regex_match(requirement_field_value):
>               raise StrictDocSemanticError.not_comma_separated_tag_field(
                    requirement_field=requirement_field,
                    **get_location(requirement),
                )
E               strictdoc.backend.sdoc.error_handling.StrictDocSemanticError: ('Requirement field of type Tag is invalid: Tag 1, Tag 2, Tag 3', "Tag field requires ', '-separated values.", 7, 1, None)

..\..\..\..\..\strictdoc\backend\sdoc\validations\requirement.py:157: StrictDocSemanticError
```

**Comment by stanislaw** [comment]
Just to make sure: you are saying this test also fails for you on the unmodified `main` branch? I was in the middle of some other work, but I can investigate. For this to happen, I would first need to reproduce it somehow.

**Comment by GGBeer** [comment]
OK, I found the culprit.
Its strictdoc/backend/sdoc/models/document_grammar.py line:86
in Main branch a `GrammarElementFieldString()` is created for the TAGS field, in my Use-Constants branch a `GrammarElementFieldTag()` is created. Must have copied the whole element instead of only the string-constant change.
The GrammarField instance type is used in the strictdoc/backend/sdoc/validations/requirement.py::validate_requirement_field(). For the Strings-Variant there is no specific validationCheck, i.e. misses the  `elif isinstance(grammar_field, GrammarElementFieldTag)` part. With my change it does enter this validationCheck part.

As I have to cleanup my branches anyway, I could either change it back to GrammarElementFieldString and we fix it later, or we decide to leave it as is and change the validationCheck for GrammarElementFieldTag fields.

Regards,
Guus

**Comment by stanislaw** [comment]
Good that there is clarity now!

> As I have to cleanup my branches anyway, I could either change it back to GrammarElementFieldString and we fix it later, 

I suggest that we go with the smallest steps. Let's change it back first in this PR and attempt to fix it in a separate one.

**Comment by GGBeer** [comment]
OK!

File: strictdoc/backend/sdoc/document_reference.py [hunk 1]
```diff
@@ -1,9 +1,15 @@
+from __future__ import annotations
+
 from typing import Optional
+from typing import TYPE_CHECKING
```
Comment (stanislaw) [inline strictdoc/backend/sdoc/document_reference.py:4]: Must this change be implemented within your overall change of introducing constants?

File: strictdoc/backend/sdoc/document_reference.py [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
 
-from strictdoc.backend.sdoc.models.document import Document
+if TYPE_CHECKING:
+    from strictdoc.backend.sdoc.models.document import Document
 
 
 class DocumentReference:
+    _document: Document | None
```
Comment (stanislaw) [inline strictdoc/backend/sdoc/document_reference.py:11]: The same as for the previous line: how changing this helps to achieve the goal of introducing the constants?

**Review by stanislaw — CHANGES_REQUESTED** [review]
We are almost there, but please address these remaining comments:

- Two remarks about `TYPE_CHECKING` which does not seem to be necessary for merging this work.
- Clean Git history. Please rebase your branch against the main branch. Please check the suggested Git workflow: https://strictdoc.readthedocs.io/en/latest/strictdoc-6-contributing.html#git-workflow:

> The branch should be always rebased against the main branch. The git fetch && git rebase origin/main is preferred over git fetch && git merge origin/main.

File: strictdoc/backend/sdoc/document_reference.py (hunk 2 above)
Comment (GGBeer) [reply strictdoc/backend/sdoc/document_reference.py:11]: By fixing the circular depency, PyCharm hinted to add a separate variable type hint. Not sure if this improves how python loads/initializes the classes.
Note: Found out that the used Union-Variant was [introduced with python 3.10](https://docs.python.org/3.10/library/typing.html#typing.Optional).
I will check if we can do without the hint or change it into Optional[Document].

File: strictdoc/backend/sdoc/document_reference.py (hunk 1 above)
Comment (GGBeer) [reply strictdoc/backend/sdoc/document_reference.py:4]: Unfortunately yes.
The constants are defined as part of the specific document node class files, e.g. ReferenceType is defined in reference.py. By referencing the constants, often circular import dependencies are introduced. In combination with type hints, this gets even worse. For example Document=>Requirement=>DocumentReference=>Document was one of them. Moving the "from import" statements into local imports is not a good solution which is also faulted by lint/black.
Found the  TYPE_CHECKING fix on [Adam Johnson's Blog](https://adamj.eu/tech/2021/05/13/python-type-hints-how-to-fix-circular-imports/).

Comment (stanislaw) [reply strictdoc/backend/sdoc/document_reference.py:4]: Ok, it looks like by adding a few more imports you are hitting the dependency cycles. I have tried your branch on my machine and it looks like moving the newly introduced `RequirementFieldName` and the already existing `RESERVED_NON_META_FIELDS` and `RequirementFieldType` to `models/type_system.py` solves the cycles problem for now.

Within your branch, just move these types to `models/type_system.py` and you will no longer have to fight with the cycle.

My context: I am aware of the dependencies problem but I would like to solve it within some other work, not such a trivial introduction of enum values. For one thing, I was going to remove the `DocumentReference` very soon.

I see that you have removed the merge commit by rebasing. Could you make the next step and squash all commits into just one? (interactive rebase:`rebase -i` procedure)

**Review by stanislaw — APPROVED** [review]

**Pull request merged**